standards/core/rln-contract.md

```diff
@@ -0,0 +1,204 @@
+---
+title: WAKU2-RLN-CONTRACT
+name: Waku2 RLN Contract Specification
+category: Standards Track
+tags: [waku/core-protocol]
+editor: Sergei Tikhomirov <sergei@status.im>
+contributors:
+---
+
+## Abstract
+
+This specification describes the smart contract that governs RLN memberships, in particular:
+
+- the overall smart contract functionality;
+- contract parameters;
+- the values of those parameters for the initial deployment;
+- the way in which the parameters can be modified.
+
+## Background
+
+Rate-Limiting Nullifiers (RLN) is a ZK-based rate-limiting technique used in Waku.
+Before sending messages, users must register in a smart contract.
+They then attach a ZK proof of valid membership to every message.
+Relaying nodes check message validity and drop invalid messages.
+
+A message is considered valid in terms of RLN if:
+- its sender has an active RLN membership; and
+- the sender hasn't exceeded their rate limit within the current epoch.
+
+RLN is managed by a smart contract that keeps track of active memberships.
+Users interact with the contract to register a new membership and
+to obtain Merkle proofs and the tree root,
+which is necessary for proof generation and verification.
+
+In testnet deployment, the RLN contract does not require any payment apart from gas costs.
+This document aims to make the RLN contract suitable for mainnet deployment.
+In particular, the contract would require uses to put down a deposit.
+The aim of the deposit initially is purely to protect the system from abuse.
+It may also be explored as a revenue stream for Waku in later versions.
+
+## Semantics
+
+The RLN contract provides the following functionality:
+- register a membership in a given tier;
+- check if a membership is valid;
+- withdraw the deposit for an expired membership;
+- extend an expired membership.
+
+For the `Owner`, the contract provides the following additional functionality:
+- change the modifiable parameters (see parameter table below);
+- (TBD) freeze certain functionality (e.g., in case of an attack, stop new registrations, deposit withdrawals, etc).
+
+In the initial deployment:
+- there is no slashing;
+- membership expiration includes a grace period;
+- a user can extend their membership without putting up any additional deposit.
+
+### Membership life-cycle
+
+To register a membership, a user puts locks up a deposit.
+Each membership has an expiration date.
+After a membership expires, it enters a grace period, during which the user can extend it.
```

Comment:
What does a user need to extend the membership? The secret or to control the wallet which executed the inital registration? Might be worth mentioning it somewhere

Reply:
Good question.

IIUC, there are three types of identifiers: RLN ID, Waku node ID, and Ethereum address from which transactions are sent to the smart contract. As established elsewhere, Waku node ID and Ethereum address are largely irrelevant in the context of this document.

I see two options regarding the RLN ID:
- only allow extending your own membership, as proven by a signature from a private key from which your RLN commitment ID is derived;
- allow anyone to extend anyone's membership.

Pros and cons of "you can only extend your own membership":
- pro: confirms to the principle of least privileges (see a possible attack described below);
- con: risks linking RLN commitment ID with the Ethereum address from which the extension transaction is sent;
- con: complicated managing many users' RLN memberships by a single service provider (I'm not sure of the exact security model here).

Pros and cons of "anyone can extend any membership":
- pro: conceptually simple, no extra check (and related gas expenses) required;
- pro: simplifies centralized membership management on behalf of many users;
- con: allows an attack where Alice wants to withdraw her deposit towards the end of the grace period, but Bob extends her membership earlier without her consent, so that Alice's deposit remains locked up for another 90 days.

@fryorcraken @jm-clius @alrevuelta @richard-ramos thoughts?

Reply:
Intuitively, if we are going to allow the redeem-my-payment scheme, we should go for a "you can only extend your own membership" for now. Allowing others to extend any membership sounds like a possible future enhancement (if it proves necessary for some use case).

Reply:
OK, let's only allow the owner of the membership to extend it.

Although this seems to link the memberhip to an Ethereum address from which the extension transaction is sent, I've realized that it's actually within our security model: IIUC, the privacy guarantee is that _messages_ are not linked to a particular membership, not that a membership is not linked to an Ethereum address.